Add Hugepage-Isolator

swatisehgal opened the pull request:
+ support for Hugetlbfs Cgroup in Kubelet

Co-Authored-By: Louise Daly louise.m.daly@intel.com

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#pull-request-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
```

ConnorDoyle reviewed (changes requested):
Could you please remove the following from the patch?

* Duplicated isolator library files (import the existing one instead?)
* The hugepage-isolator executable (also add to .gitignore in that directory)

lmdaly commented:
Hey @ConnorDoyle made those changes. Thanks for the review!

squall0gd commented on cluster/addons/hp-client/examples/example-hugepage-pod.yaml:
```diff
@@ -0,0 +1,19 @@
+spec:
+  containers:
+  - name: nginxcontainer
+    image: si-docker.ir.intel.com/ssehgal/mmap
```
Is this image is visible/available outside of Intel? As far as we want to upstream it, community won't be able to use your example ;)

squall0gd commented on cluster/addons/hp-client/examples/example-hugepage-pod.yaml:
```diff
@@ -0,0 +1,19 @@
+  - name: nginxcontainer
+    image: si-docker.ir.intel.com/ssehgal/mmap
+    command: [ "/bin/bash", "-c", "--" ]
+    args: [ "while true; do sleep 300000; done;" ]
```
You can prepare your image like this:
```docker
...
ENTRYPOINT ["/bin/bash", "-c", "--", "sleep"]
CMD "inf"
```

Your will be able run this image w/o any command and args. If you need to change amount of time, then you can do it with command, by passing time like this:

```yaml
    command: 300
```

squall0gd commented on cluster/addons/hp-client/examples/hugepage-client.yaml:
```diff
@@ -0,0 +1,19 @@
+        - -v
+        - "10"
+        - -logtostderr
+        image: si-docker.ir.intel.com/lmdaly/cgroup-hugepage-isolator
```
It's better to use `cgroup-hugepage-isolator` here(which is declared in Dockerfile)

squall0gd commented on cluster/addons/hp-client/hugepages/discovery/hp_discovery.go:
```diff
@@ -0,0 +1,22 @@
+        }
+        return hpTotalNum, nil
+}
+
```
remove this line

squall0gd commented on cluster/addons/hp-client/hugepage-isolator.go:
```diff
@@ -0,0 +1,38 @@
+	name = "cgroup-hugetlb-limit"
+)
+
+// TODO: split it to smaller functions
```
is it still valid?

squall0gd commented on cluster/addons/hp-client/hugepage-isolator.go:
```diff
@@ -0,0 +1,38 @@
+	// kubelet eventDispatcher address
+	eventDispatcherAddress = "localhost:5433"
+	// iso-client own address
+	isolatorLocalAddress = "localhost:5444"
```
Can you change your `isolatorLocalAddress` to sth different? It is overlapping with already existing `coreaffinity` isolator.

squall0gd commented on cluster/addons/hp-client/hugepages/discovery/hp_discovery.go:
```diff
@@ -0,0 +1,22 @@
+)
+
+func DiscoverHugepageNumber() (int, error) {
+        procFsPath := "/proc/sys/vm/nr_hugepages"
```
s/procFsPath/hugepageInfoPath/

squall0gd commented on cluster/addons/hp-client/hugepages/hpIsolator.go:
```diff
@@ -0,0 +1,163 @@
+		}
+		if spec.Hugepage != "" {
+			return spec.Hugepage, nil
+		}
```
you can move annotations parsing to another function, so each of your function will be responsible for one task :)

squall0gd commented on cluster/addons/hp-client/hugepages/hpIsolator.go:
```diff
@@ -0,0 +1,163 @@
+		},
+	}
+	
+	return cgroupResource, nil
```
probably you should split logic here to other functions.

swatisehgal committed:
Add Hugepage-Isolator
+ support for Hugetlbfs Cgroup in Kubelet

Co-Authored-By: Louise Daly louise.m.daly@intel.com

Commit:
Added Dockerfile & go file for sample hugepge application
Moved annotation parsing from getHugeSize() to getAnnotationFromPod()
Moved map creation to getHugepageMap() and IsolationControl Value formatting to getCgroupResourceValue
Provided appropriate image names in Pod Manifests
Changed local isolator address to fix conflict with Coreaffinity isolator

Co-Authored-By: Swati Sehgal swati.sehgal@intel.com
